lightning/src/ln/peer_handler.rs

```diff
@@ -1332,27 +1370,59 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 	/// [`ChannelManager::process_pending_htlc_forwards`]: crate::ln::channelmanager::ChannelManager::process_pending_htlc_forwards
 	/// [`send_data`]: SocketDescriptor::send_data
 	pub fn process_events(&self) {
+		let mut _single_processor_lock = self.event_processing_lock.try_lock();
+		if _single_processor_lock.is_err() {
+			// While we could wake the older sleeper here with a CV and make more even waiting
+			// times, that would be a lot of overengineering for a simple "reduce total waiter
+			// count" goal.
+			match self.blocked_event_processors.compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire) {
+				Err(val) => {
+					debug_assert!(val, "compare_exchange failed spuriously?");
+					return;
+				},
+				Ok(val) => {
+					debug_assert!(!val, "compare_exchange succeeded spuriously?");
+					// We're the only waiter, as the running process_events may have emptied the
+					// pending events "long" ago and there are new events for us to process, wait until
+					// its done and process any leftover events before returning.
+					_single_processor_lock = Ok(self.event_processing_lock.lock().unwrap());
+					self.blocked_event_processors.store(false, Ordering::Release);
+				}
+			}
+		}
+
+		let mut disconnect_peers = HashMap::new();
+		let mut events_generated = self.message_handler.chan_handler.get_and_clear_pending_msg_events();
+		events_generated.append(&mut self.message_handler.route_handler.get_and_clear_pending_msg_events());
+
 		{
 			// TODO: There are some DoS attacks here where you can flood someone's outbound send
 			// buffer by doing things like announcing channels on another node. We should be willing to
 			// drop optional-ish messages when send buffers get full!
 
-			let mut peers_lock = self.peers.lock().unwrap();
-			let mut events_generated = self.message_handler.chan_handler.get_and_clear_pending_msg_events();
-			events_generated.append(&mut self.message_handler.route_handler.get_and_clear_pending_msg_events());
-			let peers = &mut *peers_lock;
+			let peers_lock = self.peers.read().unwrap();
+			let peers = &*peers_lock;
 			macro_rules! get_peer_for_forwarding {
```

Comment:
does this need to be a macro? Returning a type suggesting whether or not a loop should be continued seems fairly straightforward.

Reply:
Hm, I think the match can be pretty short, possibly even a one-liner, but it's not a big deal.